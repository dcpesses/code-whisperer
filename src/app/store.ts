import {configureStore, ThunkAction, Action} from '@reduxjs/toolkit';
import counterReducer from '@/features/redux-counter/counterSlice';
import modalCommandListReducer from '@/features/modal-command-list/modalSlice';
// import twitchReducer from '../features/twitch/twitchSlice';
import channelReducer from '@/features/twitch/channel-slice.js';
import userReducer from '@/features/player-queue/user-slice.js';
import queueReducer from '@/features/player-queue/queue-slice';
import todosReducer from '@/features/todos/todosSlice';

const reducer = {
  counter: counterReducer,
  channel: channelReducer,
  modal: modalCommandListReducer,
  queue: queueReducer,
  user: userReducer,
  todos: todosReducer,
};
export const store = configureStore({ reducer });

export function getStoreWithState(preloadedState?: RootState) {
  return configureStore({ reducer, preloadedState });
}

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
