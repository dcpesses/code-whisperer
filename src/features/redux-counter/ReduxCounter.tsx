import { useState } from 'react';

import { useAppSelector, useAppDispatch } from '../../app/hooks';
import {
  decrement,
  increment,
  incrementByAmount,
  incrementAsync,
  incrementIfOdd,
  selectCount,
} from './counterSlice';
import styles from './ReduxCounter.module.css';

export function ReduxCounter() {
  const count = useAppSelector(selectCount);
  const dispatch = useAppDispatch();
  const [incrementAmount, setIncrementAmount] = useState('2');

  const incrementValue = Number(incrementAmount) || 0;

  return (
    <div>
      <div className={styles.row}>
        <button
          className={styles.button}
          aria-label="Decrement value"
          onClick={() => dispatch(decrement())}
        > - </button>
        <span
          className={styles.value}
          aria-label="Counter value">{count}</span>
        <button
          className={styles.button}
          aria-label="Increment value"
          onClick={() => dispatch(increment())}
        > + </button>
      </div>
      <div className={styles.row}>
        <input className={styles.textbox}
          aria-label="Set increment amount"
          value={incrementAmount}
          onChange={(e) => setIncrementAmount(e.target.value)}
        />
        <button className={styles.button}
          aria-label="Add increment amount"
          onClick={() => dispatch(incrementByAmount(incrementValue))}
        >
          Add
        </button>
        <div>
          <button
            className={styles.asyncButton}
            aria-label="Add async increment amount"
            onClick={() => dispatch(incrementAsync(incrementValue))}
          >
            Add Async
          </button>
          <button
            className={styles.button}
            aria-label="Add increment amount if odd"
            onClick={() => dispatch(incrementIfOdd(incrementValue))}
          >
            Add If Odd
          </button>
        </div>

      </div>
    </div>
  );
}
