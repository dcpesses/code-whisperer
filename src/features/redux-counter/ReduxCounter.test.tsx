import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import * as counterAPI from './counterAPI';

import { getStoreWithState } from '../../app/store';
import { Provider } from 'react-redux';
import { ReduxCounter } from './ReduxCounter';
import { Store, UnknownAction } from '@reduxjs/toolkit';

describe('ReduxCounter', () => {
  let store: Store<unknown, UnknownAction, unknown>;
  beforeEach(() => {
    store = getStoreWithState();
  });
  test('Should render as expected', () => {
    const {container} = render(
      <Provider store={store}>
        <ReduxCounter />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });
  test('Should render and display state counter value', () => {
    render(
      <Provider store={store}>
        <ReduxCounter />
      </Provider>
    );
    const counterElement = screen.getByLabelText('Counter value', {selector: 'span'});

    expect(counterElement).toHaveTextContent('0');
    expect(counterElement).toMatchSnapshot();
  });

  test('Should increment and display updated counter value', () => {
    render(
      <Provider store={store}>
        <ReduxCounter />
      </Provider>
    );
    const counterElement = () => screen.getByLabelText('Counter value', {selector: 'span'});
    const button = screen.getByLabelText('Increment value');

    expect(counterElement()).toHaveTextContent('0');

    fireEvent.click(button);

    expect(counterElement()).toHaveTextContent('1');
  });

  test('Should decrement and display updated counter value', () => {
    render(
      <Provider store={store}>
        <ReduxCounter />
      </Provider>
    );
    const counterElement = () => screen.getByLabelText('Counter value', {selector: 'span'});
    const button = screen.getByLabelText('Decrement value');

    expect(counterElement()).toHaveTextContent('0');

    fireEvent.click(button);

    expect(counterElement()).toHaveTextContent('-1');
  });

  test('Should increment counter using numeric input amount', () => {
    render(
      <Provider store={store}>
        <ReduxCounter />
      </Provider>
    );
    const counterElement = () => screen.getByLabelText('Counter value', {selector: 'span'});
    const button = screen.getByLabelText('Add increment amount');

    expect(counterElement()).toHaveTextContent('0');

    const input = screen.getByRole('textbox');
    expect(input).toHaveValue('2');
    fireEvent.change(input, {target: {value: '3'}});
    expect(input).toHaveValue('3');

    fireEvent.click(button);

    expect(counterElement()).toHaveTextContent('3');

  });
  test('Should async increment counter using numeric input amount', async() => {
    render(
      <Provider store={store}>
        <ReduxCounter />
      </Provider>
    );
    const counterElement = () => screen.getByLabelText('Counter value', {selector: 'span'});
    expect(counterElement()).toHaveTextContent('0');
    expect(screen.getByRole('textbox')).toHaveValue('2');

    const button = screen.getByLabelText('Add async increment amount');

    fireEvent.click(button);

    await waitFor(() => expect(counterElement()).toHaveTextContent('2'));
  });
  test('Should handle async error when incrementing counter using numeric input amount', async() => {
    vi.spyOn(counterAPI, 'fetchCount').mockRejectedValue({});
    render(
      <Provider store={store}>
        <ReduxCounter />
      </Provider>
    );
    const counterElement = () => screen.getByLabelText('Counter value', {selector: 'span'});
    const button = screen.getByLabelText('Add async increment amount');

    fireEvent.click(button);

    await waitFor(() => expect(counterElement()).toHaveTextContent('0'));
  });
  test('Should increment counter only when counter value is odd using numeric input amount', () => {
    render(
      <Provider store={store}>
        <ReduxCounter />
      </Provider>
    );
    const counterElement = () => screen.getByLabelText('Counter value', {selector: 'span'});
    expect(counterElement()).toHaveTextContent('0');
    expect(screen.getByRole('textbox')).toHaveValue('2');

    // confirm counter will not increment
    const button = screen.getByLabelText('Add increment amount if odd');
    fireEvent.click(button);
    expect(counterElement()).toHaveTextContent('0');

    // set counter to odd value
    fireEvent.click(screen.getByLabelText('Increment value'));
    expect(counterElement()).toHaveTextContent('1');

    // confirm counter increments
    fireEvent.click(button);
    expect(counterElement()).toHaveTextContent('3');
  });
  test('Should handle invalid input values', () => {
    render(
      <Provider store={store}>
        <ReduxCounter />
      </Provider>
    );
    const counterElement = () => screen.getByLabelText('Counter value', {selector: 'span'});
    const button = screen.getByLabelText('Add increment amount');

    expect(counterElement()).toHaveTextContent('0');

    const input = screen.getByRole('textbox');
    fireEvent.change(input, {target: {value: 'f'}});
    fireEvent.click(button);

    expect(counterElement()).toHaveTextContent('0');
  });
});

