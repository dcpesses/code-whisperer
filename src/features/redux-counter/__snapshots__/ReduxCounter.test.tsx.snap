// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ReduxCounter > Should render and display state counter value 1`] = `
<span
  aria-label="Counter value"
  class="_value_503e3a"
>
  0
</span>
`;

exports[`ReduxCounter > Should render as expected 1`] = `
<div>
  <div>
    <div
      class="_row_503e3a"
    >
      <button
        aria-label="Decrement value"
        class="_button_503e3a"
      >
         - 
      </button>
      <span
        aria-label="Counter value"
        class="_value_503e3a"
      >
        0
      </span>
      <button
        aria-label="Increment value"
        class="_button_503e3a"
      >
         + 
      </button>
    </div>
    <div
      class="_row_503e3a"
    >
      <input
        aria-label="Set increment amount"
        class="_textbox_503e3a"
        value="2"
      />
      <button
        aria-label="Add increment amount"
        class="_button_503e3a"
      >
        Add
      </button>
      <div>
        <button
          aria-label="Add async increment amount"
          class="_asyncButton_503e3a"
        >
          Add Async
        </button>
        <button
          aria-label="Add increment amount if odd"
          class="_button_503e3a"
        >
          Add If Odd
        </button>
      </div>
    </div>
  </div>
</div>
`;
